[doc] PR #2 for Document Partitions in Explorer Section Doc7178 (#7601)
yugabyte/yugabyte-db: docs/content/latest/explore/ysql-language-features/partitions.md

## Overview

> Partitioning is another term for physically dividing large tables in YugabyteDB into smaller, more manageable tables to improve performance. With regards to the columns, the ones containing timestamps are usually subject to partitioning because of the historical and predictable nature of their data.

Author: Rephrased

> Since partitioned tables do not appear nor act differently from the original table, applications accessing the database are not always aware of the fact that partitioning has taken place.

Author: Thanks for the suggestion.

YSQL supports the following types of partitioning:

- Range partitioning, when a table is partitioned into ranges defined by one or more key columns. In this case, the ranges of values assigned to partitions do not overlap.
- List partitioning, when a table is partitioned via listing key values to appear in each partition.
- Hash partitioning, when a table is partitioned by specifying a modulus and remainder for each  partition.

## Declarative Table Partitioning

YSQL allows you to specify how exactly to divide a table. You provide a partitioning method and partition key consisting of a list of columns or expressions. The divided table is called a partitioned table, and the resulting tables are called partitions. When you insert rows into a partitioned table, they are redirected to a partition depending on the value of the partition key. You can also directly insert rows into the partition table itself, and those rows can be fetched by querying the parent table.

You can nest partitions, in which case they woud have their own distinct indexes, constraints, and default values.

> A regular table cannot become a partitioned table, just as a partitioned table cannot become a regular table. That said, YSQL allows attaching a regular table (provided it has the same schema as that of the partitioned table) as a partition to the partitioned  table. Conversely, a partition can be detached from the partitioned table, in which case it can behave as a regular table that is not part of the partitioning hierarchy.

Author: Thanks for the suggestion.

> A partitioned table and its partitions have hierarchical structure and are subject to its rules, except the following:

Author: Changed.

> A partitioned table and its partitions have hierarchical structure and are subject to its rules, except the following:

Reviewer: ```suggestion
Partitions are subject to their parent table's rules, with the following exceptions:
```

- If there are no partitions, you can use the `ONLY` clause to add or drop a constraint on the partitioned table. If there are partitions, you cannot use `ONLY` because adding or deleting constraints on the  partitioned table is not supported in such cases.

> - Partitions cannot have columns which do not exist in the partitioned table. You can add a partition using `ALTER TABLE ... ATTACH PARTITION` if their columns match the partitioned table, including a `oid` column.

Author: Changed.

> - Partitions inherit `CHECK` and `NOT NULL` constraints of their partitioned table. This does not include `CHECK` constraints marked `NO INHERIT`.

Reviewer 2: You cannot create constraints marked NO INHERIT at all. Please check the documentation of Postgres 11.2 before adding statements here. It's also very important to add test cases (or log GH issues requesting for tests to be added) for all statements mentioned in the docs, otherwise it is very embarrassing for the company to document a behavior that is not supported. 

"yugabyte=# alter table plt1_e add constraint cons check (a > 0) no inherit;
ERROR:  cannot add NO INHERIT constraint to partitioned table "plt1_e"

Author: Correct, and that's what we have in the doc. Perhaps the sentence wasn't as clear as it could have been, so I modified it a little:

A partitioned table and its partitions have hierarchical structure and are subject to its rules, except the following:
- If there are no partitions, you can use the `ONLY` clause to add or drop a constraint on the partitioned table. If there are partitions, you cannot use `ONLY` because adding or deleting constraints on the  partitioned table is not supported in such cases.
- You can add a regular table to the partition hierarchy using `ALTER TABLE ... ATTACH PARTITION` if their columns match the partitioned table, including a `oid` column.
- Partitions inherit `CHECK` and `NOT NULL` constraints of the partitioned table, with the following exceptions: 
  - `CHECK` constraints marked `NO INHERIT`be cannot be created on partitioned tables.
  - The `NOT NULL` constraint cannot be added to a partition's column if the constraint is present in their partitioned table.

> - The `NOT NULL` constraint cannot be added to a partition's column if the constraint is present in their partitioned table.

Reviewer 2: Why would one even add a constraint that's already present in the partitioned table? Did you mean "cannot be dropped" instead of "cannot be added"?

Author: Yes. Here's what we have:

A partitioned table and its partitions have hierarchical structure and are subject to its rules, except the following:

- If there are no partitions, you can use the `ONLY` clause to add or drop a constraint on the partitioned table. If there are partitions, you cannot use `ONLY` because adding or deleting constraints on the  partitioned table is not supported in such cases.
- You can add a regular table to the partition hierarchy using `ALTER TABLE ... ATTACH PARTITION` if their columns match the partitioned table, including a `oid` column.
- Partitions inherit `CHECK` and `NOT NULL` constraints of the partitioned table, with the following exceptions: 
  - `CHECK` constraints marked `NO INHERIT`be cannot be created on partitioned tables.
  - The `NOT NULL` constraint cannot be dropped on a partition's column if the constraint is present in their partitioned table.

YSQL allows you to specify how exactly to divide a table. You provide a partitioning method and partition key consisting of a list of columns or expressions. The divided table is called a partitioned table, and the resulting tables are called partitions. When you insert rows into a partitioned table, they are redirected to a partition depending on the value of the partition key. You can also directly insert rows into the partition table itself, and those rows can be fetched by querying the parent table.

You can nest partitions, in which case they woud have their own distinct indexes, constraints, and default values.

A regular table cannot become a partitioned table, just as a partitioned table cannot become a regular table. That said, YSQL allows attaching a regular table (provided it has the same schema as that of the partitioned table) as a partition to the partitioned  table. Conversely, a partition can be detached from the partitioned table, in which case it can behave as a regular table that is not part of the partitioning hierarchy.

A partitioned table and its partitions have hierarchical structure and are subject to its rules, except the following:

> - If there are no partitions, you can use the `ONLY` clause to add or drop a constraint on the partitioned table. If there are partitions, you cannot use `ONLY` because adding or deleting constraints on the  partitioned table is not supported in such cases.

Reviewer 2: This statement is wrong. Please remove this. You don't need "ONLY" clause to create constraint on partitioned table or on the partitions.

Author: I misunderstood a sentence in the postgresql doc regarding inheritance exceptions. 
I removed the bulleted list and here's what we have now:

_A regular table cannot become a partitioned table, just as a partitioned table cannot become a regular table. That said, YSQL allows attaching a regular table (provided it has the same schema as that of the partitioned table) as a partition to the partitioned  table. Conversely, a partition can be detached from the partitioned table, in which case it can behave as a regular table that is not part of the partitioning hierarchy. A partitioned table and its partitions have hierarchical structure and are subject to most of its rules._

- You can add a regular table to the partition hierarchy using `ALTER TABLE ... ATTACH PARTITION` if their columns match the partitioned table, including a `oid` column.
- Partitions inherit `CHECK` and `NOT NULL` constraints of the partitioned table, with the following exceptions: 

>   - `CHECK` constraints marked `NO INHERIT`be cannot be created on partitioned tables.

Reviewer 2: Neither of the following two statements look like exceptions?!

Author: I misunderstood a sentence in the postgresql doc regarding inheritance exceptions.
I removed the bulleted list and here's what we have now:

A regular table cannot become a partitioned table, just as a partitioned table cannot become a regular table. That said, YSQL allows attaching a regular table (provided it has the same schema as that of the partitioned table) as a partition to the partitioned table. Conversely, a partition can be detached from the partitioned table, in which case it can behave as a regular table that is not part of the partitioning hierarchy. A partitioned table and its partitions have hierarchical structure and are subject to most of its rules.

> Since partitioned tables do not appear nor act differently from the original table, applications accessing the database are not always aware of the fact that partitioning has taken place.

Author: Removed the "Typically..." sentence

YSQL supports the following types of partitioning:

> - Range partitioning, when a table is partitioned into ranges defined by one or more key columns. In this case, the ranges of values assigned to partitions do not overlap.

Reviewer 2: Actually we should specify HASH partitioning here, like PostgreSQL does, maybe we can skip examples and all, but its good to mention it here. Otherwise it gives the impression that although we are using a fork of Postgres we are not supporting hash partitioning. 
https://www.postgresql.org/docs/11/ddl-partitioning.html

Author: Added hash to the list

- List partitioning, when a table is partitioned via listing key values to appear in each partition.

## Declarative Table Partitioning

> YSQL allows you to specify how exactly to divide a table. You provide a partitioning method and partition key consiting of a list of columns or expressions. The divided table is called a partitioned table, and the resulting tables are called partitions. When you insert rows into a partitioned table, they are redirected to a partition depending on the value of the partition key.

Reviewer 2: Please add at the end of this paragraph: You can also directly insert rows into the partition table itself, and those rows can be fetched by querying the parent table.

Author: Added this sentence. Here's what we have now:
YSQL allows you to specify how exactly to divide a table. You provide a partitioning method and partition key consisting of a list of columns or expressions. The divided table is called a partitioned table, and the resulting tables are called partitions. When you insert rows into a partitioned table, they are redirected to a partition depending on the value of the partition key. You can also directly insert rows into the partition table itself, and those rows can be fetched by querying the parent table.

You can nest partitions, in which case they woud have their own distinct indexes, constraints, and default values.

> A regular table cannot become a partitioned table, just as a partitioned table cannot become a regular table. That said, YSQL supports adding a regular table that has the same schema as that of the partitioned table to be attached as one of the partitions of the partitioned table. Conversely, a partition can be detached from the partitioned table, in which case it can behave as a regular table that is not part of the partitioning hierarchy.

Reviewer 2: My bad, apologies. Can you change the sentence "YSQL supports adding a regular table" into "YSQL allows attaching a regular table (provided it has the same schema as that of the partitioned table) as a partition to the partitioned table.

Author: Changed

);
```

> *change_date* represents the date when any type of change occured on the employee record. This date is typically required when generating monthly reports and only the last month's data is queried. To reduce the amount of old data that  needs to be stored, you can reduce the amount of data to store and keep these data for one year, and on the first of every month you remove the data from the previous month. You can partition the `employees` table to meet these requirements. You start by specifying bounds corresponding to the partitioning method and partition key of the `employees` table. This means you create partitions as regular tables and YSQL generates partition constraints automatically based on the partition bound specification every time they have to be referenced.

Reviewer 2: *change_date* represents the date when any type of change occured on the employee record. This date is typically required when generating monthly reports. Let us assume that typically only the last month's data is queried often, and so every month a periodic job cleans up the data in the table that is older than a year.
To make this workflow easier, you can partition the `employees` table. You start by specifying bounds...

Author: Modified this paragraph

);
```

> *change_date* represents the date when any type of change occurred on the employee record. This date is typically required when generating monthly reports. Assuming that typically only the last month's data is queried often, then the data that is older than one year is removed from the table every month. To simplify this process, you can partition the `employees` table. You start by specifying bounds corresponding to the partitioning method and partition key of the `employees` table. This means you create partitions as regular tables and YSQL generates partition constraints automatically based on the partition bound specification every time they have to be referenced.

Reviewer 2: My bad again. 
"Assume that typically only the last month's data is queried often and any data older than a year is removed from the table periodically".

Author: Thanks for your suggestion.

DROP TABLE employees_2019_02;
```

> The following example shows how to remove the partition from the partitioned table while retaining access to it as a separate table which enables you to performed operations on the data:

Reviewer 2: performed operations --> perform operations

Author: fixed

The following example shows how to remove the partition from the partitioned table while retaining access to it as a separate table which enables you to performed operations on the data:

```sql
ALTER TABLE employees DETACH PARTITION employees_2019_02;
```

> The following example shows how to add a new partition to deal with new data by creating an empty partition in the partitioned table:

Reviewer 2: Oh you can add back this section. Please only remove the TABLESPACE clause in the CREATE statements.

Author: Added

> DROP TABLE employees_2019_02;

Reviewer 2: Above: "(you need to take an `ACCESS EXCLUSIVE` lock on the partitioned `employees` table)"

this statement makes it seem like a user needs to do something to achieve this DROP statement. The right statement here is "Note that this operation takes an ACCESS EXCLUSIVE lock on the partitioned employees table"

Author: Changed.

- If you choose to define row triggers, you do so on individual partitions instead of the partitioned table.
- You cannot mix temporary and permanent relations in the same partition hierarchy.

> ## Constraint Exclusion

Reviewer 2: This section is nowhere as clear as the equivalent in Postgres documentation unfortunately: https://www.postgresql.org/docs/11/ddl-partitioning.html#DDL-PARTITIONING-CONSTRAINT-EXCLUSION

Also weirdly why have we chosen this section (which is more applicable for legacy inheritance) rather than partition pruning? Why have we also not chosen the section on Partitioning Best Practices?

Author: Our doc site is not structured like postgres docs. Our Explore section contains overview-type articles only. 
Thanks for your input.
